pkg/gadgets/traceloop/tracer/tracer.go

```diff
@@ -0,0 +1,549 @@
+
+	// 1. Create inner Map as perf buffer.
+	innerBuffer, err := ebpf.NewMap(innerBufferSpec)
+	if err != nil {
+		return fmt.Errorf("error creating inner map: %w", err)
+	}
+
+	// 2. Use this inner Map to create the perf reader.
+	perfReader, err := perf.NewReaderWithOptions(innerBuffer, gadgets.PerfBufferPages*os.Getpagesize(), perf.ReaderOptions{WriteBackward: true, OverWritable: true})
+	if err != nil {
+		innerBuffer.Close()
+
+		return fmt.Errorf("error creating perf ring buffer: %w", err)
+	}
+
+	// 3. Add the inner map's file descriptor to outer map.
+	t.objsLock.Lock()
+	err = t.objs.MapOfPerfBuffers.Put(mntnsID, innerBuffer)
+	t.objsLock.Unlock()
+	if err != nil {
+		innerBuffer.Close()
+		perfReader.Close()
+
+		return fmt.Errorf("error adding perf buffer to map with mntnsID %d: %w", mntnsID, err)
+	}
+
+	t.readers.Store(containerID, &reader{
+		perfReader:      perfReader,
+		previousHeadPos: make([]uint64, getRingsNumber(perfReader)),
+		mntnsID:         mntnsID,
+	})
+
+	return nil
+}
+
+func (t *Tracer) Read(containerID string) ([]*types.Event, error) {
+	syscallContinuedEventsMap := make(map[uint64][]*syscallEventContinued)
+	syscallEnterEventsMap := make(map[uint64][]*syscallEvent)
+	syscallExitEventsMap := make(map[uint64][]*syscallEvent)
+	events := make([]*types.Event, 0)
+
+	r, ok := t.readers.Load(containerID)
+	if !ok {
+		return nil, fmt.Errorf("no perf reader for %q", containerID)
+	}
+
+	reader, ok := r.(*reader)
+	if !ok {
+		return nil, errors.New("the map should only contain *reader")
+	}
+
+	if reader.perfReader == nil {
+		log.Infof("reader for %v is nil, it was surely detached", containerID)
+
+		return nil, nil
+	}
+
+	err := readOverWritable(reader, func(record perf.Record, size uint32) error {
+		var cSyscallEvent C.struct_syscall_event_t
+		var cSyscallContEvent C.struct_syscall_event_cont_t
+
+		switch uintptr(size) {
+		case alignSize(unsafe.Sizeof(cSyscallEvent)):
+			cSyscallEvent = *(*C.struct_syscall_event_t)(unsafe.Pointer(&record.RawSample[0]))
+
+			event := &syscallEvent{
+				timestamp: uint64(cSyscallEvent.timestamp),
+				typ:       uint8(cSyscallEvent.typ),
+				contNr:    uint8(cSyscallEvent.cont_nr),
+				cpu:       uint16(cSyscallEvent.cpu),
+				id:        uint16(cSyscallEvent.id),
+				pid:       uint32(cSyscallEvent.pid),
+				comm:      C.GoString(&cSyscallEvent.comm[0]),
+				mountNsID: reader.mntnsID,
+			}
+
+			var typeMap *map[uint64][]*syscallEvent
+			switch event.typ {
+			case syscallEventTypeEnter:
+				event.args = make([]uint64, syscallArgs)
+				for i := uint8(0); i < syscallArgs; i++ {
+					event.args[i] = uint64(cSyscallEvent.args[i])
+				}
+
+				typeMap = &syscallEnterEventsMap
+			case syscallEventTypeExit:
+				// In the C structure, args is an array of uint64.
+				// But in this particular case, we used it to store a C long, i.e. the
+				// syscall return value, so it is safe to cast it to golang int.
+				event.retval = int(cSyscallEvent.args[0])
```

Comment:
Confusing... on my machine:
* C.int is 4 bytes
* C.long is 8 bytes
* But Golang's int is 8 bytes

```
$ cat sizeof.c 
#include <stdio.h>
int main() {
	printf("int: %d\n", sizeof(int));
	printf("long: %d\n", sizeof(long));
	return 0;
}
$ ./sizeof 
int: 4
long: 8
```

https://go.dev/play/p/5sx0KOxCp35

So it seems to work. But why making it explicit by defining the retval field as a int64?